src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -135,8 +135,16 @@ enum ResolutionFailure<'a> {
     ///
     /// `Namespace` is the namespace specified with a disambiguator
     /// (as opposed to the actual namespace of the `Res`).
-    WrongNamespace(Res, /* disambiguated */ Namespace),
-    /// The link failed to resolve. `resolution_failure` should look to see if there's
+    WrongNamespace {
+        /// What the link resolved to.
+        res: Res,
+        /// The expected namespace for the resolution, determined from the link's disambiguator.
+        ///
+        /// E.g., for `[fn@Result]` this is [`Namespace::ValueNS`],
+        /// even though `Result`'s actual namespace is [`Namespace::TypeNS`].
+        expected_ns: Namespace,
+    },
```

Comment:
"expected" is a bad name for this IMO. I would call it "disambiguated_ns" or something like that.

Reply:
But I feel that "expected" makes it clear that it's not what the actual namespace is — just like "expected" vs. "found" in rustc errors.

Reply:
"expected" is ambiguous whether you expect it to be the actual value or whether you expect it to be the disambiguated namespace.

Reply:
> expect it to be the actual namespace

I would expect it to be the expected namespace ;)

"expected" usually means what should be the case, not necessarily what is the case, as opposed to "actual", which means what actually is the case. If you can come up with something clearer, I'll go with that, but `disambiguated_ns` seems confusing to me.

Reply:
Well, I don't know that I agree, but either way having the doc-comments seems useful, so this is strictly better than the previous code even if it's still not perfectly clear.